From 6a1b5962ab182df03e89d115ba0548f72012d2fa Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Wed, 29 Oct 2014 00:31:18 -0400
Subject: [drm] qxl: don't create too large primary surface

Message-id: <1414542678-17321-1-git-send-email-airlied@redhat.com>
Patchwork-id: 98501
O-Subject: [PATCH] [PATCH RHEL7.1 BZ1158233] qxl: don't create too large primary surface
Bugzilla: 1158233
RH-Acked-by: Rob Clark <rclark@redhat.com>

From: Marc-Andre Lureau <marcandre.lureau@gmail.com>

bugzilla: 1158233
upstream: c572aaf46f71f63ae5914d4e194a955e0ba1b519

Limit primary to qemu vgamem size, to avoid reaching
qemu guest bug "requested primary larger than framebuffer"
on resizing screen too large to fit.

Remove unneeded and misleading variables.

Related to:
https://bugzilla.redhat.com/show_bug.cgi?id=1127552

Signed-off-by: Marc-Andre Lureau <marcandre.lureau@redhat.com>
Cc: stable@vger.kernel.org
Signed-off-by: Dave Airlie <airlied@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/qxl/qxl_display.c b/drivers/gpu/drm/qxl/qxl_display.c
index 5d7ea24..98a344c 100644
--- a/drivers/gpu/drm/qxl/qxl_display.c
+++ b/drivers/gpu/drm/qxl/qxl_display.c
@@ -523,7 +523,6 @@ static int qxl_crtc_mode_set(struct drm_crtc *crtc,
  struct qxl_framebuffer *qfb;
  struct qxl_bo *bo, *old_bo = NULL;
  struct qxl_crtc *qcrtc = to_qxl_crtc(crtc);
- uint32_t width, height, base_offset;
  bool recreate_primary = false;
  int ret;
  int surf_id;
@@ -553,9 +552,10 @@ static int qxl_crtc_mode_set(struct drm_crtc *crtc,
  if (qcrtc->index == 0)
   recreate_primary = true;
 
- width = mode->hdisplay;
- height = mode->vdisplay;
- base_offset = 0;
+ if (bo->surf.stride * bo->surf.height > qdev->vram_size) {
+  DRM_ERROR("Mode doesn't fit in vram size (vgamem)");
+  return -EINVAL;
+        }
 
  ret = qxl_bo_reserve(bo, false);
  if (ret != 0)
@@ -569,10 +569,10 @@ static int qxl_crtc_mode_set(struct drm_crtc *crtc,
  if (recreate_primary) {
   qxl_io_destroy_primary(qdev);
   qxl_io_log(qdev,
-      "recreate primary: %dx%d (was %dx%d,%d,%d)\n",
-      width, height, bo->surf.width,
-      bo->surf.height, bo->surf.stride, bo->surf.format);
-  qxl_io_create_primary(qdev, base_offset, bo);
+      "recreate primary: %dx%d,%d,%d\n",
+      bo->surf.width, bo->surf.height,
+      bo->surf.stride, bo->surf.format);
+  qxl_io_create_primary(qdev, 0, bo);
   bo->is_primary = true;
  }
 
-- 
1.7.1